docs(connection-router):  add documentation for connection router (#212)
axelarnetwork/axelar-amplifier: doc/src/contracts/connection_router.md

# Connection Router Contract

The connection router contract is responsible for routing messages to a chain and controlling gateways associated with a chain that is connected to Axelar.

> 

Reviewer: This would be a useful place to add some comments about the governance and admin accounts (what each of them can do).

## Interface

```Rust
pub enum ExecuteMsg {
    /*
     * Governance Methods
     * All of the below messages can only be called by governance
     */
    // Registers a new chain with the router
    RegisterChain {
        chain: ChainName,
        gateway_address: String,
    },
    // Changes the gateway address associated with a particular chain
    UpgradeGateway {
        chain: ChainName,
        contract_address: String,
    },

    /*
     * Router Admin Methods
     * All of the below messages can only be called by the router admin
     */
    // Freezes a chain, in the specified direction.
    FreezeChain {
        chain: ChainName,
        direction: GatewayDirection,
    },
    // Unfreezes a chain, in the specified direction.
    UnfreezeChain {
        chain: ChainName,
        direction: GatewayDirection,
    },

    /*
     * Gateway Messages
     * The below messages can only be called by registered gateways
     */

>     // Routes a message to all outgoing gateways registered to the destination domain.

Reviewer: ```suggestion
    // Routes each message to the gateway registered to the destination chain.
```

Can you update the comment in the contract's msg.rs file too, if need be?

    // Called by an incoming gateway
    RouteMessages(Vec<Message>),
}

#[derive(QueryResponses)]
pub enum QueryMsg {}
```

> ## Events

Reviewer: I don't think we need this section. This is a bit too low level for this doc, and these events are not mentioned outside of this section.

```Rust
pub struct RouterInstantiated {
    pub admin: Addr,
    pub governance: Addr,
    pub nexus_gateway: Addr,
}

pub struct ChainRegistered {
    pub name: ChainName,
    pub gateway: Addr,
}

pub struct GatewayInfo {
    pub chain: ChainName,
    pub gateway_address: Addr,
}

pub struct GatewayUpgraded {
    pub gateway: GatewayInfo,
}

pub struct GatewayFrozen {
    pub gateway: GatewayInfo,
}

pub struct GatewayUnfrozen {
    pub gateway: GatewayInfo,
}

pub struct ChainFrozen {
    pub name: ChainName,
}

pub struct ChainUnfrozen {
    pub name: ChainName,
}

pub struct MessageRouted {
    pub msg: Message,
}
```

## Connection Router graph

```mermaid
graph TD

e[External Gateway]

> t[Gateway]

Reviewer: This should be in the subgraph, since the gateway is deployed on axelar

subgraph Axelar
g[Governance]
a[Admin]
c[Connection Router]
end

g--RegisterChain-->c
g--UpgradeGateway-->c
a--FreezeChain-->c
a--UnfreezeChain-->c
e-.->t
t--RouteMessages-->c

> ```

Reviewer: For this diagram, I think it would be more useful to just illustrate the message flow, and leave all of the governance and admin stuff out. I would like to see multiple gateways though, to show a message flowing from one gateway to another.

We also should show the relayer, relaying the message from the external gateway to the gateway on axelar.

## Message Routing sequence diagram

```mermaid
sequenceDiagram
autonumber
participant External Gateway A
box LightYellow Axelar
participant Gateway A
participant Connection Router
participant Gateway B
end
participant External Gateway B

External Gateway A->>+Gateway A: 
Gateway A->>+Connection Router: ExecuteMsg::RouteMessages
Connection Router->>+Gateway B: ExecuteMsg::RouteMessages
Gateway B->>+External Gateway B: 
```

> 1. The External Gateway sends an incoming message to Gateway

Reviewer: The external gateway doesn't send the message. The external gateway emits an event, and then the relayer relays that event from the external gateway to the gateway on axelar. Let's include the relayer in this diagram.

> 2. Gateway receives the incoming messages and calls Connection Router to route it to the destination Gateway.

Reviewer: ```suggestion
2. Gateway receives the incoming messages, verifies the messages, and then passes the messages to the Connection Router to route them to the destination Gateway.
```

3. The Connection Router sends outgoing messages to the destination Gateways.

> 4. The destination Gateway emits a MessageRouted event, and the message goes to the destination's External Gateway

Reviewer: The router emits `MessageRouted`. Also, the message does not immediately go to the external gateway. It usually needs to be included in a proof, and then the relayer relays it.

    pub gateway: GatewayInfo,
}

pub struct GatewayUnfrozen {
    pub gateway: GatewayInfo,
}

pub struct ChainFrozen {
    pub name: ChainName,
}

pub struct ChainUnfrozen {
    pub name: ChainName,
}

pub struct MessageRouted {
    pub msg: Message,
}
```

## Connection Router graph

```mermaid
graph TD

ea[External Gateway A]
ra[Relayer A]
subgraph Axelar
ta[Gateway A]
c[Connection Router]
tb[Gateway B]
m[Multisig Prover]
end
rb[Relayer B]
eb[External Gateway B]

ea--emit event-->ra
ra--RouteMessages-->ta
ta--RouteMessages-->c
c--RouteMessages-->tb
tb--get messages-->m
m--construct proof-->rb
rb--send proof-->eb

```

## Message Routing sequence diagram

```mermaid
sequenceDiagram
autonumber
participant External Gateway A
participant Relayer A
box LightYellow Axelar
participant Gateway A
participant Connection Router
participant Gateway B
participant Multisig Prover
end
participant Relayer B
participant External Gateway B

External Gateway A->>+Relayer A: event emitted
Relayer A->>+Gateway A: gateway::ExecuteMsg::RouteMessages
Gateway A->>+Connection Router: connection_router::ExecuteMsg::RouteMessages
Connection Router->>+Gateway B: gateway::ExecuteMsg::RouteMessages

> Gateway B-->>-Connection Router: GatewayEvent::MessageRouted

Reviewer: The event is not sent to the router. It's picked up by the relayer. I would just remove all of the mentions of the events in this diagram, for clarity

## Interface

```Rust
pub enum ExecuteMsg {
    /*
     * Governance Methods
     * All of the below messages can only be called by governance
     */
    // Registers a new chain with the router
    RegisterChain {
        chain: ChainName,
        gateway_address: String,
    },
    // Changes the gateway address associated with a particular chain
    UpgradeGateway {
        chain: ChainName,
        contract_address: String,
    },

    /*
     * Router Admin Methods
     * All of the below messages can only be called by the router admin
     */
    // Freezes a chain, in the specified direction.
    FreezeChain {
        chain: ChainName,
        direction: GatewayDirection,
    },
    // Unfreezes a chain, in the specified direction.
    UnfreezeChain {
        chain: ChainName,
        direction: GatewayDirection,
    },

    /*
     * Gateway Messages
     * The below messages can only be called by registered gateways
     */
    // Routes each message to the gateway registered to the destination chain.
    // Called by a registered gateway
    RouteMessages(Vec<Message>),
}
```

> ## Events

Reviewer: Remove the header
